ui/perfherder/CompareSubtestDistributionView.jsx

```diff
@@ -0,0 +1,376 @@
+    const {
+      originalProject: originalProjectName,
+      originalSubtestSignature,
+      originalRevision,
+      newRevision,
+      newSubtestSignature,
+      newProject: newProjectName,
+    } = this.props.$stateParams;
+    const { originalProject, newProject } = await this.fetchProjectsToCompare(
+      originalProjectName,
+      newProjectName,
+    );
+
+    const {
+      testSuite,
+      subtest,
+      testName,
+      platform,
+    } = await this.fetchTestCharacteristics(
+      originalProjectName,
+      originalRevision,
+      originalSubtestSignature,
+    );
+    this.setState({
+      testSuite,
+      subtest,
+      testName,
+      platform,
+      originalProject,
+      newProject,
+    });
+    const syncPromises = this.syncGraphData(
+      originalProject,
+      originalRevision,
+      originalSubtestSignature,
+      newProject,
+      newRevision,
+      newSubtestSignature,
+    );
+
+    Promise.all(syncPromises).then(() => {
+      window.document.title = `${this.state.platform}: ${this.state.testName}`;
+      this.setState({ dataLoading: false });
+    });
+  }
+
+  syncProjectGraphData = async (
+    projectName,
+    revision,
+    subtestSignature,
+    withBaseline,
+  ) => {
+    const {
+      graphSpecsField,
+      resultSetField,
+      replicateErrorField,
+      drawingDataField,
+      targetId,
+    } = withBaseline
+      ? {
+          graphSpecsField: 'graphOneSpecs',
+          resultSetField: 'originalResultSet',
+          replicateErrorField: 'originalReplicateError',
+          drawingDataField: 'baseDrawingData',
+          targetId: 'Base',
+        }
+      : {
+          graphSpecsField: 'graphTwoSpecs',
+          resultSetField: 'newResultSet',
+          replicateErrorField: 'newReplicateError',
+          drawingDataField: 'newDrawingData',
+          targetId: 'New',
+        };
+
+    const { replicateData, drawingData } = await this.fetchReplicateGraphData(
+      projectName,
+      revision,
+      subtestSignature,
+    );
+    const { resultSet, replicateDataError } = replicateData;
+    const numRuns = drawingData ? drawingData.numRuns : 0;
+
+    return this.setState({
+      [resultSetField]: resultSet,
+      [replicateErrorField]: replicateDataError,
+      [drawingDataField]: drawingData,
+      [graphSpecsField]: this.getGraphSpecs(
+        targetId,
+        numRuns,
+        replicateDataError,
+      ),
+    });
+  };
+
+  syncGraphData = (
+    originalProject,
+    originalRevision,
+    originalSubtestSignature,
+    newProject,
+    newRevision,
+    newSubtestSignature,
+  ) => {
+    const originalSyncPromise = this.syncProjectGraphData(
+      originalProject.name,
+      originalRevision,
+      originalSubtestSignature,
+      true,
+    );
+    const newSyncPromise = this.syncProjectGraphData(
+      newProject.name,
+      newRevision,
+      newSubtestSignature,
+      false,
+    );
+    return [originalSyncPromise, newSyncPromise];
```

Comment:
Anytime you see repeated logic, there is an opportunity to create a new component. Since both of the graphs use the same state (`newProject, newRevision` vs `originalProject, originalRevision`) and have the same behavior this would be a good use case. It'll make this component easier to read, easier to make changes that apply to both and less heavy during the `componentDidMount` cycle. So while this component could be responsible for verifying query parameters, etc, the new component can fetch data and render the data for each graph. You can see how I approached that with the compareChooser [here](https://github.com/mozilla/treeherder/blob/master/ui/perfherder/SelectorCard.jsx) - the two cards share the same logic so they have a separate component that is used in the top level [compareChooser] view component(https://github.com/mozilla/treeherder/blob/master/ui/perfherder/CompareSelectorView.jsx). What do you think about trying this approach?